arch/risc-v/src/bl602/bl602_spi.c

```diff
@@ -0,0 +1,1232 @@
+ *   Initialize bl602 SPI hardware interface
+ *
+ * Input Parameters:
+ *   dev      - Device-specific state data
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void bl602_spi_init(FAR struct spi_dev_s *dev)
+{
+  FAR struct bl602_spi_priv_s *priv = (FAR struct bl602_spi_priv_s *)dev;
+  const struct bl602_spi_config_s *config = priv->config;
+
+  /* Initialize the SPI semaphore that enforces mutually exclusive access */
+
+  nxsem_init(&priv->exclsem, 0, 1);
+
+  bl602_configgpio(BOARD_SPI_CS);
+  bl602_configgpio(BOARD_SPI_MOSI);
+  bl602_configgpio(BOARD_SPI_MISO);
+  bl602_configgpio(BOARD_SPI_CLK);
+
+  /* set master mode */
+
+  bl602_set_spi_0_act_mode_sel(1);
+
+  /* spi cfg  reg:
+   * cr_spi_deg_en 1
+   * cr_spi_m_cont_en 0
+   * cr_spi_byte_inv 0
+   * cr_spi_bit_inv 0
+   */
+
+  modifyreg32(BL602_SPI_CFG, 0, SPI_CFG_CR_DEG_EN);
+  modifyreg32(BL602_SPI_CFG, SPI_CFG_CR_M_CONT_EN, 0);
+  modifyreg32(BL602_SPI_CFG, SPI_CFG_CR_BYTE_INV, 0);
+  modifyreg32(BL602_SPI_CFG, SPI_CFG_CR_BIT_INV, 0);
+
+  /* disable rx ignore */
+
+  modifyreg32(BL602_SPI_CFG, SPI_CFG_CR_RXD_IGNR_EN, 0);
+
+  bl602_spi_setfrequency(dev, config->clk_freq);
+  bl602_spi_setbits(dev, 8);
+  bl602_spi_setmode(dev, config->mode);
+
+  /* spi fifo clear */
+
+  modifyreg32(BL602_SPI_CFG, SPI_FIFO_CFG_0_RX_CLR
+              | SPI_FIFO_CFG_0_TX_CLR, 0);
+}
+
+/****************************************************************************
+ * Name: bl602_spi_deinit
+ *
+ * Description:
+ *   Deinitialize bl602 SPI hardware interface
+ *
+ * Input Parameters:
+ *   dev      - Device-specific state data
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void bl602_spi_deinit(FAR struct spi_dev_s *dev)
+{
+  FAR struct bl602_spi_priv_s *priv = (FAR struct bl602_spi_priv_s *)dev;
+
+  bl602_swrst_ahb_slave1(AHB_SLAVE1_SPI);
+
+  priv->frequency = 0;
+  priv->actual = 0;
+  priv->mode = SPIDEV_MODE0;
+  priv->nbits = 0;
+  priv->dma_chan = 0;
+}
+
+/****************************************************************************
+ * Name: bl602_spibus_initialize
+ *
+ * Description:
+ *   Initialize the selected SPI bus
+ *
+ * Input Parameters:
+ *   Port number (for hardware that has multiple SPI interfaces)
+ *
+ * Returned Value:
+ *   Valid SPI device structure reference on success; a NULL on failure
+ *
+ ****************************************************************************/
+
+FAR struct spi_dev_s *bl602_spibus_initialize(int port)
+{
+  FAR struct spi_dev_s *spi_dev;
+  FAR struct bl602_spi_priv_s *priv;
+  irqstate_t flags;
+
+  switch (port)
+    {
+#ifdef CONFIG_BL602_SPI0
+    case 0:
+      priv = &bl602_spi_priv;
+      break;
+#endif
+  default:
+    return NULL;
+    }
+
+  spi_dev = (FAR struct spi_dev_s *)priv;
+
+  flags = enter_critical_section();
+
+  if ((volatile int)priv->refs != 0)
```

Comment:
Why does this need to be cast volatile?  I don't think hardware should be touching this address.